Mathlib/LinearAlgebra/Matrix/DFT.lean

```diff
@@ -0,0 +1,274 @@
+  by_contra hg
+  rw [Complex.exp_eq_one_iff] at hg
+  obtain ⟨z, hz⟩ := hg
+  rw [mul_div_assoc, mul_comm (z:ℂ) _, (mul_right_inj' two_pi_I_ne_zero),
+      (div_eq_iff_mul_eq (Nat.cast_ne_zero.2 (NeZero.ne _)))] at hz
+  norm_cast at hz
+  apply_fun ( Int.mod · m) at hz
+  simp only [Int.mul_mod_left, Int.subNatNat_eq_coe] at hz
+  by_cases h1 : p ≤ k
+  · rw [Int.mod_eq_of_lt, eq_comm, sub_eq_zero] at hz
+    norm_cast at hz
+    apply h (Fin.ext hz)
+    simp only [sub_nonneg, Nat.cast_le, Fin.val_fin_le]
+    apply (h1)
+    rw [← Nat.cast_sub]
+    norm_cast
+    apply Nat.sub_lt_right_of_lt_add
+    exact_mod_cast h1
+    apply (Nat.lt_add_right _ _ _ (Fin.is_lt _))
+    exact_mod_cast h1
+  · push_neg at h1
+    rw [ ← neg_sub, mod_eq_mod_neg, eq_comm, neg_eq_zero] at hz
+    rw [Int.mod_eq_of_lt, sub_eq_zero, eq_comm] at hz
+    norm_cast at hz
+    apply h (Fin.ext hz)
+    simp only [neg_sub, sub_nonneg, Nat.cast_le, Fin.val_fin_le]
+    apply le_of_lt
+    apply (h1)
+    rw [← Nat.cast_sub]
+    norm_cast
+    apply Nat.sub_lt_right_of_lt_add
+    apply le_of_lt
+    exact_mod_cast h1
+    apply (Nat.lt_add_right _ _ _ (Fin.is_lt _))
+    apply le_of_lt h1
+
+/-- The IDFT Matrix Invertible "def/instance"-/
+noncomputable def invWₙ : Invertible (Wₙ n) := by
+  apply invertibleOfRightInverse  _ (of (fun (k p : Fin n) => exp (-2 * π * I * k * p / n) / n)) _
+  funext k p
+  -- rw [mul_apply]
+  simp_rw [mul_apply, Wₙ_apply, of_apply, mul_div_assoc', ← Complex.exp_add]
+  by_cases h : k = p
+  · rw [h, one_apply_eq]
+    ring_nf
+    simp_rw [Complex.exp_zero, Finset.sum_const, card_fin, nsmul_eq_mul]
+    apply mul_inv_cancel_left₀ (Nat.cast_ne_zero.2 (NeZero.ne _))
+  · rw [one_apply_ne h]
+    have h1 : ∀ (x : Fin n), 2 * π * I * k * x / n +  (-2) * π * I * x * p / n  =
+        (x : ℕ) * ((2*π * I * (k - p)) / n) := by intro x; ring
+    simp_rw [h1, ← sum_div, div_eq_zero_iff, Complex.exp_nat_mul, Fin.sum_univ_eq_sum_range]
+    left
+    rw [geom_sum_eq, ← Complex.exp_nat_mul, div_eq_zero_iff]
+    left
+    rw [sub_eq_zero, Complex.exp_eq_one_iff]
+    use (k - p)
+    rw [mul_div_cancel', mul_comm, mul_eq_mul_right_iff]
+    left
+    simp_rw [Int.cast_sub, Int.cast_ofNat]
+    exact (Nat.cast_ne_zero.2 (NeZero.ne _))
+    apply cexp_sub_ne_one _ _ h
+
+/-- The IDFT Matrix -/
+theorem dftMatrix_inv  [Invertible (Wₙ n)] :
+    ⅟(Wₙ n) = of (fun (k p : Fin n) => exp ( -2 * π * I * k * p / n) / n) := by
+  letI := (invWₙ n)
+  convert (rfl : ⅟(Wₙ n) = _)
+
+lemma iWₙ_apply (k p : Fin n) : (Wₙ n)⁻¹ k p = exp (-2 * π * I * k * p / n)/n := by
+  letI := invWₙ n
+  rw [← Matrix.invOf_eq_nonsing_inv (Wₙ n), dftMatrix_inv, of_apply]
+
+/-- The DFT operation and DFT matrix applied to sequence/vector are tha same -/
+theorem dft_eq_Wₙ_mul (v : Fin n → ℂ) : dft n v = mulVec (Wₙ n) v := by
+  funext r
+  simp only [dft, mulVec, dotProduct, Wₙ_apply]
+
+/-- The IDFT operation and IDFT matrix applied to sequence/vector are tha same -/
```

Comment:
```suggestion
/-- The IDFT operation and IDFT matrix applied to sequence/vector are the same -/
```

I think you've gone overboard on some docstrings and not enough on more complicated ones, or the ones that set up the definitions. Some theorems are fine without docstrings.